cmd/govim/highlight.go

```diff
@@ -0,0 +1,80 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/govim/govim/cmd/govim/internal/types"
+)
+
+type propDict struct {
+	Highlight string `json:"highlight"`
+	Combine   bool   `json:"combine,omitempty"`
+	Priority  int    `json:"priority,omitempty"`
+	StartIncl bool   `json:"start_incl,omitempty"`
+	EndIncl   bool   `json:"end_incl,omitempty"`
+}
+
+func (v *vimstate) textpropDefine() error {
+	v.BatchStart()
+	for _, s := range []types.Severity{types.SeverityErr, types.SeverityWarn, types.SeverityInfo, types.SeverityHint} {
+		hi := types.SeverityHighlight[s]
+
+		// Note that we reuse the highlight name as text property name, even if they aren't the same thing.
+		v.BatchChannelCall("prop_type_add", hi, propDict{
+			Highlight: hi,
+			Combine:   true, // Combine with syntax highlight
+			Priority:  types.SeverityPriority[s],
+		})
+	}
+	res := v.BatchEnd()
+	for i := range res {
+		if v.ParseInt(res[i]) != 0 {
+			return fmt.Errorf("call to prop_type_add() failed")
+		}
+	}
+	return nil
+}
+
+func (v *vimstate) redefineHighlights(diags []types.Diagnostic) error {
+	v.BatchStart()
+	defer v.BatchEnd()
```

Comment:
You want `v.BatchCancelIfNotEnded()` in this defer and should instead explicitly call `v.BatchEnd()` on the happy path below